[Step3] 자동차 경주

yjna2316 opened the pull request:
안녕하세요~ 

제가 작성한 코드 중 추가 설명하고 싶은 내용들은 커맨트로 달아 놓았어요!

yjna2316 committed:
feat: UI 관련 로직 InputView, ResultView 추가
  - InputView: 입력값 읽어오기 & 예외처리
  - ResultView: 결과 출력

yjna2316 committed:
feat: Car, CarRacing 클래스 및 자동차 경기 구현

yjna2316 committed:
README.md와 테스트코드 추가

yjna2316 committed:
chore: ktlint 수정

yjna2316 commented on src/main/kotlin/racingcar/view/InputView.kt:
```diff
@@ -0,0 +1,18 @@
+package racingcar.view
+
```
UI 관련 로직과 비즈니스 로직 분리 위해 view, domain 패키지로 분리해놓았어요.

yjna2316 commented on src/main/kotlin/racingcar/domain/CarRacing.kt:
```diff
@@ -0,0 +1,16 @@
+    }
+
+    private fun randomNumber() = (0..9).random()
+}
```
0과 9사이 무작위 값을 정하는것은 경주가 가진 Rule이라 생각해서, 자동차가 아닌 자동차 경주의 관심사로 생각해 여기에 추가해놓았어요.!

malibinYun replied:
Car의 관심사가 아니라고 생각이 드셔서 자동차 경주에 해당 로직을 위치시키셔도, 여전히 CarRacing 객체는 랜덤의 요소 때문에 테스트하기 어려운 구조입니다.
랜덤 숫자를 생성하는 인터페이스를 만들고, 해당 구현체를 외부로부터 주입 받게끔 만들어 보는 것은 어떨까요?

yjna2316 replied:
@malibinYun 
```
object RandomGenerator {
    fun randomNumber() = (0..9).random()
}
```

저는 랜덤 숫자를 생성하는 RandomGenerator 클래스를 위처럼 따로 분리했는데,  
`랜덤 숫자를 생성하는 인터페이스를 만들고, 해당 구현체를 외부로부터 주입 받게끔 만들어 보는 것은 어떨까요?`.말씀 주신 내용은 
인터페이스를 만들어서, 인터페이스를 구현체로 구현하는 방법을 말씀하시는 걸까용?

예) 
```
interface RandomGenerator {
    fun randomNumber(): Int
}

object randomCreator: RandomGenerator {
    override fun randomNumber(): Int = (0..9).random()
}
```

malibinYun replied:
네, CarRacing 객체를 만들 때 RandomGenerator를 파라미터로 받는다면, 원하는 구현체를 주입함으로써 테스트 가능한 형태가 됩니다 :)

yjna2316 commented on src/main/kotlin/racingcar/domain/CarRacing.kt:
```diff
@@ -0,0 +1,16 @@
+    fun moveCars(listCar: List<Car>): List<Car> {
+        return listCar.map { car ->
+            val nextPosition = car.move(randomNumber())
+            Car(position = nextPosition)
```
data class로 Car를 정의할 경우 copy메소드로 새로운 객체를 만들 수 있을 텐데 (아래처럼), 코드가 지저분해보이는 이유로 Car를  class로 정의했어요.

data class는 layer간 데이터 전달할때 dto로 주로 사용하는 것으로 알고 있는데, 이 상황을 제외하고 보통 어떤 상황에서 data class와 class를 사용하는지 궁금해요! 

```
fun moveCars(listCar: List<Car>): List<Car> {
        return listCar.map { car ->
            val nextPosition = car.move(randomNumber())
            car.copy(position = nextPosition)
}
```

malibinYun replied:
data class는 단순 POJO class로 사용하기 용이하지만, DTO의 용도로 만들어진 키워드가 아닙니다.
data class와 일반 class의 차이점은 무엇일까요?

yjna2316 replied:
data class 는 equals, hashcode, toString이 자동 생성되고 class 는 그렇지 않다 라는 차이점이 있어요!

yjna2316 committed:
불필요한Readme 내용 제거

malibinYun commented on src/main/kotlin/racingcar/README.md:
```diff
@@ -0,0 +1,41 @@
+# [3단계] 초간단 자동차 경주 구현
+
+# 요구사항
+- [X] 주어진 횟수 동안 n대의 자동차는 전진 또는 멈출 수 있다.
+- [X] 사용자는 몇 대의 자동차로 몇 번의 이동을 할 것인지를 입력할 수 있어야 한다.
+- [x] 전진하는 조건은 0에서 9 사이에서 무작위 값을 구한 후 무작위 값이 4 이상일 경우이다.
+- [X] 자동차의 상태를 화면에 출력한다. 어느 시점에 출력할 것인지에 대한 제약은 없다.
+- [X] 핵심 로직을 구현하는 코드와 UI를 담당하는 로직을 구분한다.
+    - View와 Domain 패키지로 분리
+    - View - InputView, ResultView와 같은 UI 클래스
+    - Domain - 자동차 경주 핵심 로직
+- [X] 모든 로직에 단위 테스트를 구현한다. 단, UI(System.out, System.in) 로직은 제외
```
요구 사항 정리 아주 좋습니다 👍👍

malibinYun commented on src/main/kotlin/racingcar/domain/Car.kt:
```diff
@@ -0,0 +1,16 @@
+    fun move(number: Int): Int {
+        val moveOneStepOrNot = if (number >= LOWER_BOUND_TO_MOVE) MOVE else STOP
+        return position.plus(moveOneStepOrNot)
+    }
```
파라미터로 넘어온 `number` 의 이름만으로는 무엇을 의미하는 지 알기 어렵습니다.
더 구체적인 이름으로 지어보는 것은 어떨까요?

또, move의 반환 값으로 position의 결과가 나올 것이라고 예상하기 어렵습니다.
Integer, 정수라는 것 만으로 position을 유추하기는 어렵기 때문이에요. 
move의 반환값으로 새로운 Car를 반환해보는 것은 어떨까요? 또는, DIstance라는 래퍼 클래스를 더 만들 수 있겠네요🙂

malibinYun commented on src/test/kotlin/racingcar/domain/CarFixture.kt:
```diff
@@ -0,0 +1,16 @@
+    )
+
+    val car = Car(position = 0)
+}
```
테스트에 사용하는 사용하는 로직은 모두 하드 코딩하는 것이 바람직합니다. CarFixture처럼 공통된 부분을 만들어 여러 테스트에서 사용하면, 각 테스트들이 서로 결합되어버리는 상황이 발생합니다.
테스트를 진행하다가 CarFixture 내부가 변경된다면, 해당 객체를 사용하고 있던 모든 테스트들을 다시 검증해봐야합니다.
또, 다른 종류의 테스트를 만들어야 한다면 또 새로운 픽스처를 만들어야하는 상황이 발생합니다.

해당 내용을 잘 설명한 블로그 글을 공유드립니다 :)[https://jojoldu.tistory.com/611](https://jojoldu.tistory.com/611)

malibinYun commented on src/test/kotlin/racingcar/domain/CarRacingTest.kt:
```diff
@@ -0,0 +1,28 @@
+        resultCars[0].position shouldBeInRange runningCars[0].position..runningCars[0].position.plus(1)
+        resultCars[1].position shouldBeInRange runningCars[1].position..runningCars[1].position.plus(1)
+        resultCars[2].position shouldBeInRange runningCars[2].position..runningCars[2].position.plus(1)
+    }
```
범위를 테스트 하는 것은 좋은 테스트 방법이 아닙니다.
리팩터링을 진행하다가 프로덕션 코드에서 실수로 범위 이외의 값을 반환하게 만들었다 해도, 랜덤의 요소 때문에 틀린 코드가 테스트 통과하는 경우가 발생합니다.

랜덤 숫자를 반환하는 대한 인터페이스를 만들었다면, 고정 값을 반환하는 구현체를 하드코딩으로 주입함으로써 프로그래머가 숫자를 제어할 수 있게 됩니다.
의존성 주입을 사용해서 리팩터링 해보는 것은 어떨까요? 🙂

malibinYun commented on src/main/kotlin/RacingCarApplication.kt:
```diff
@@ -0,0 +1,19 @@
+    (0 until trialCount).map {
+        listCar = moveCars(listCar)
+        ResultView.show(listCar)
+    }
```
moveCars는 도메인 로직입니다.
main 함수에 도메인 로직과 뷰 로직이 결합되어있네요.
이 자동차 경주 프로젝트를 그대로 웹 또는 모바일에 이식해도, 도메인 로직에는 문제가 없게끔 만드는 것을 목표로 해보는 것을 추천드립니다 🙂
자동차를 움직이고 결과를 반환하게 하는 로직을 더 추상화 해보는 것은 어떨까요?

malibinYun reviewed (changes requested):
안녕하세요 :) 3단계 미션 고생 많으셨습니다.

생각해보면 좋을 점들에 대해 코멘트 남겨두었습니다.
저번 단계의 피드백 반영이 없어, 이전 미션 피드백 반영도 부탁드립니다. 🙂
3, 4단계 PR을 동시에 올려주셨는데요, 3단계에서 많은 변화가 있을 것으로 판단되어 3단계를 우선적으로 완료해주시고, 그 이후 4단계 리뷰를 진행하겠습니다.

너무 급하게 진행하지 않아도 괜찮습니다. 천천히 미션 내용을 곱씹으시면서 충분한 생각과 함께 미션을 수행하시길 권장드립니다 :)

피드백 반영 후 다시 리뷰 요청 주세요!

yjna2316 committed:
refac: 코드리뷰 반영 - 난수 생성하는 RandomUtil 클래스 생성

yjna2316 committed:
feat: README 업데이트

yjna2316 committed:
refac: 리뷰반영 - main 함수에 도메인 로직이 보이지 않도록 추상화, CarRace, RaceResult, Cars

- Cars 일급 컬렉션 생성
- RaceResult 클래스 생성 - 경기 결과
- RaceCount, TrialCount value 클래스 생성

yjna2316 committed:
refac: RandomGenerator Util 삭제하고 MoveStrategy를 moving 패키지 아래로 이동

yjna2316 committed:
refac: 리뷰 반영 - Car 클래스 - 이동 정책 moveStrategy를 외부에서 주입 받도록 수정

yjna2316 committed:
refac: car 패키지 추가 - car 관련 도메인 이동

yjna2316 committed:
refac: 리뷰 반영 - 불필요한 CarFixture 제거

yjna2316 committed:
refac: Car 클래스 리팩토링 - 자동차 위치 의미하는 Position, Positions 추가, MoveStrategy를 Car 생성시 받도록 수정함

yjna2316 committed:
refac: Cars 클래스는 자동차 그룹을 움직이게 한다

yjna2316 committed:
test: 중복되는 코드 제거 위해 MoveStrategyFixture 생성

yjna2316 committed:
feat: 매 경기별 자동차 위치를 기억하는 RaceRecord 추가

yjna2316 committed:
refac: 자동차 경기를 시작하고 그 결과를 반환하는 CarRace 리팩토링

yjna2316 committed:
refac: ResultView 경기결과 출력하는 로직 수정

yjna2316 commented:
@malibinYun  안녕하세요~ 구조 변경해서 다시 리팩토링해보았어요.

리뷰는 [refac: 코드리뷰 반영 - 난수 생성하는 RandomUtil 클래스 생성](https://github.com/next-step/kotlin-racingcar/pull/569/commits/88252973b2a58b6dc31a6997c7a1e749a119ec0a) 여기서 부터 해주시면 되어요

malibinYun commented on src/main/kotlin/racingcar/domain/car/Car.kt:
```diff
@@ -0,0 +1,21 @@
+import racingcar.domain.moving.MoveStrategy
+import racingcar.domain.position.Position
+
+class Car(
+    private val moveStrategy: MoveStrategy
+) {
```
움직임에 대한 랜덤 요소를 외부로부터 주입 받을 수 있게 되었네요! 👍

malibinYun commented on src/main/kotlin/racingcar/domain/car/Car.kt:
```diff
@@ -0,0 +1,21 @@
+class Car(
+    private val moveStrategy: MoveStrategy
+) {
+    private var _position = Position(0)
+
+    val position: Position
+        get() = _position
```
position을 생성자로부터 주입받게 만들어보는 것은 어떨까요?
Car를 생성할 때 초기 위치를 손쉽게 조정해줄 수 있는 장점을 지닙니다.

malibinYun replied:
또한, position을 내부에서는 변형할 수 있고 외부에서는 변형할 수 없게 만들고 싶다면 set만 private로 선언해주시면 됩니다. position객체가 불변 객체이기 때문에 이런 작성이 가능합니다🙂
```suggestion
    var position = Position(0)
    private set
```

yjna2316 replied:
오 고민되는 지점이었는데, 완전 좋은 방법이네요 :) 감사합니다!

malibinYun commented on src/main/kotlin/racingcar/domain/record/RaceRecord.kt:
```diff
@@ -0,0 +1,23 @@
+    val record: List<Positions>
+        get() = _record.toList()
+
+    fun record(cars: Cars) =
+        _record.add(
+            Positions(
+                cars.cars.map(Car::position)
+            )
+        )
```
기록을 Cars를 받아 생성하는 것 보다, Cars가 직접 생성해내게 하는 것은 어떨까요?
car에 메시지를 보냄으로써 cars.cars... 같은 로직이 줄어들 것으로 기대됩니다 :)

malibinYun commented on src/test/kotlin/racingcar/domain/car/CarsTest.kt:
```diff
@@ -0,0 +1,50 @@
+import racingcar.domain.notMovingStrategy
+
+class CarsTest : AnnotationSpec() {
+    private val movingCars = List(3) {
+        Car(
+            movingStrategy
+        )
+    }
+
+    private val notMovingCars = List(3) {
+        Car(
+            notMovingStrategy
+        )
+    }
+
```
테스트에 사용하는 사용하는 로직은 모두 하드 코딩하는 것이 바람직합니다. movingCars, notMovingCars처럼 공통된 부분을 만들어 여러 테스트에서 사용하면, 각 테스트들이 서로 결합되어버리는 상황이 발생합니다.
테스트를 진행하다가 픽스처 객체의 내부가 변경된다면, 해당 객체를 사용하고 있던 모든 테스트들을 다시 검증해봐야합니다.
또, 다른 종류의 테스트를 만들어야 한다면 또 새로운 픽스처를 만들어야하는 상황이 발생합니다.

해당 내용을 잘 설명한 블로그 글을 공유드립니다 :)[https://jojoldu.tistory.com/611](https://jojoldu.tistory.com/611)

malibinYun commented on src/main/kotlin/racingcar/domain/race/CarRace.kt:
```diff
@@ -0,0 +1,43 @@
+            initCars(carCount), trialCount
+        )
+
+    private fun initCars(carCount: CarCount): Cars =
+        Cars(
+            (0 until carCount.value).map {
+                Car(RandomBasedMoveStrategy())
+            }
+        )
```
RandomBasedMoveStrategy는 매 Car를 만들 때마다 생성될 필요는 없어보입니다.
하나만 만들어두고 재사용하는 것은 어떨까요?

malibinYun replied:
또, 코틀린에서 제공해주는 함수들을 활용하면 아래처럼 작성해볼 수 있습니다 :)
```suggestion
    private fun initCars(carCount: CarCount): Cars =
        Cars(List(carCount.value) {Car(...) })
```

malibinYun reviewed (approved):
3단계 미션 고생 많으셨습니다!
바쁘신 와중에도 꾸준히 미션을 진행하시는 게 너무 좋습니다 🙂
다음 미션을 위해 머지하겠습니다. 피드백은 다음 미션에 반영해주세요!
